### PR TITLE
Make orchestrator generic and non-blocking

### PR DESCRIPTION
**Pull request by ericropiak**
This pr introduces 2 changes:

**1) Makes the action invocations non-blocking.** 
After we trigger an action, OpenWhisk returns us the activationId associated with that action. From that point on, we continually (every 5 seconds) query the OpenWhisk server to see if that action has completed. This will prevent HTTP requests from hanging and timing out while waiting for actions to finish.

**2) Makes the orchestrator generic**
Add a config.js file to contain all code specific to the pi algorithm. Now you can run `node orchestrator.js ./config.js` and the orchestrator will pull in the pi code from the config file in the same directory. This allows a user to run whatever they want using our orchestrator, as long as their config file meets the following requirements:
- Exports a function called 'action' to be used as the action run on OpenWhisk (i.e. `exports.action = someFunc()`
- Exports a function called 'reduce' to handle the results that return from OpenWhisk (i.e. `exports.reduce = someOtherFunc()`
- Exports a dictionary called 'config' with specific OpenWhisk configurations 
- Can also export a function called 'argsForAction'. This function generates arguments to be passed to an action which will be run on OpenWhisk. The function takes in an argument, which represents which number action (0 -> `NUM_ACTIONS -1`) arguments are being generated for. This allows a user to pass different arguments to different actions, if they would like

### CODE REVIEW

**Comment by danmcp** [comment]
@ericropiak Please squash the commits.

File: config.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		var totalInCircle = 0;
+		for (var i = 0; i< result.length; i++){
+			var actionResult = result[i];
+			console.log("Action " + (i + 1) + " finished with " + JSON.stringify(actionResult));
+			totalInCircle += actionResult['inCircle'];
+		}
+		console.log(totalInCircle + " out of " + TOTAL_POINTS + " were in the circle.");
+		console.log("Computed Value of Pi: " + 4 * (totalInCircle / TOTAL_POINTS));
+	} else {
+		console.log('it didnt work');
```
Comment (danmcp) [inline config.js]: Be consistent on case and you have a punctuation issue:

It didn't work

File: config.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+function computePi(result){
+	if (result){
+		var totalInCircle = 0;
+		for (var i = 0; i< result.length; i++){
+			var actionResult = result[i];
+			console.log("Action " + (i + 1) + " finished with " + JSON.stringify(actionResult));
+			totalInCircle += actionResult['inCircle'];
+		}
+		console.log(totalInCircle + " out of " + TOTAL_POINTS + " were in the circle.");
+		console.log("Computed Value of Pi: " + 4 * (totalInCircle / TOTAL_POINTS));
```
Comment (danmcp) [inline config.js]: Should be a lower case v in value.

File: config.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	// if (actionNum % 5 == 0 ){
+	// 	args['shouldFail'] = true;
+	// }
+	return args;
+}
+
+function computePi(result){
+	if (result){
+		var totalInCircle = 0;
+		for (var i = 0; i< result.length; i++){
```
Comment (danmcp) [inline config.js]: Should be a space between i and <

File: config.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+const NUM_ACTIONS = 8000;
+const TOTAL_POINTS = 100000;
+const POINTS_PER_ACTION = TOTAL_POINTS / NUM_ACTIONS;
+
+var configs = {
+	"namespace": "_", 				// Namespace to be used on OpenWhisk
```
Comment (danmcp) [inline config.js]: Comments on the right should align

File: config.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		"memory": 128,				// Max memory for each action container (gb, default 256)
+		"logs": 10					// Max memory for each action log (gb, default 10)
+	}
+}
+
+function main(params) { 
+	if (params.shouldFail) { throw 'I am failing'; }
+	var inCircle = 0;
+	var pointsPerAction = parseInt(params['pointsPerAction'])
+	for(var i=0; i<pointsPerAction; i++){
```
Comment (danmcp) [inline config.js]: be consistent on spacing in for loop syntax

File: config.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		console.log(totalInCircle + " out of " + TOTAL_POINTS + " were in the circle.");
+		console.log("Computed Value of Pi: " + 4 * (totalInCircle / TOTAL_POINTS));
+	} else {
+		console.log('it didnt work');
+	}
+};
+
+exports.configs = configs;				// Configuration values for the action on OpenWhisk
+exports.action = main;					// Function to be registered as action on OpenWhisk
+exports.argsForAction = argsForAction;	// Function to generate agrs to be passed to action
```
Comment (danmcp) [inline config.js]: Comments on the right should line up.

File: orchestrator.js [hunk 7]
```diff
@@ -1,59 +1,69 @@
 const startingMilliseconds = Date.now();
 
+// Checks if the given resource is provided, otherwise prints the given
+// error message indicating the the process failed due to an error
```
Comment (danmcp) [inline orchestrator.js]: the the

File: orchestrator.js [hunk 8]
```diff
@@ -67,74 +77,119 @@ const registerActionPromise = new Promise((resolve, reject) => {
...
+				reject(response);
+		    }
+		}
+
+		request(options, callback);
+	});
+}
+
+// Generates a promise which will query the OpenWhisk server to get the result of 
+// of the action with the given activationId. If the aciton has not finished,
```
Comment (danmcp) [inline orchestrator.js]: aciton->action

File: orchestrator.js [hunk 9]
```diff
@@ -67,74 +77,119 @@ const registerActionPromise = new Promise((resolve, reject) => {
...
-			"numPoints": POINTS_PER_ACTION
-		});
+		// Generate parameters for the action based on which action this is (0 -> NUM_ACTIONS - 1)
+		const actionArgs = custom.argsForAction ? custom.argsForAction(actionNum) : {};
+		const dataString = JSON.stringify(actionArgs);
 
-		var options = {
-		    url: 'https://' + AUTH + '@' + APIHOST + '/api/v1/namespaces/_/actions/testPoints?blocking=true',
+		const options = {
+		    url: `https://${AUTH}@${APIHOST}/api/v1/namespaces/${custom.configs['namespace']}/actions/${custom.configs['actionName']}`,
```
Comment (danmcp) [inline orchestrator.js]: It looks like you could use a helper method for forming most of these urls

**Comment by ericropiak** [comment]
All set @danmcp

File: config.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+const NUM_ACTIONS = 5000;
+const TOTAL_POINTS = 100000;
+const POINTS_PER_ACTION = TOTAL_POINTS / NUM_ACTIONS;
+
+var configs = {
+	"namespace": "_", 		// Namespace to be used on OpenWhisk
```
Comment (ravisantoshgudimetla) [inline config.js]: Isn't it namespace used by OpenWhisk on OpenShift?

File: config.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+		if (distFromCenter <= 1){
+			inCircle = inCircle + 1;
+		}
+	}
+	return {inCircle: inCircle};
+};
+
+function argsForAction(actionNum){
+	args = {pointsPerAction: POINTS_PER_ACTION}
+	// if (actionNum % 5 == 0 ){
```
Comment (ravisantoshgudimetla) [inline config.js]: Can you remove these comments, if they are unnecessary?

**Review by ravisantoshgudimetla — COMMENTED** [review]
Mostly LGTM except for nits.

File: config.js (hunk 10 above)
Comment (ericropiak) [reply config.js]: Oh yes, I'll make this more clear.

File: config.js (hunk 11 above)
Comment (ericropiak) [reply config.js]: Removed

**Pull request merged**